library/psa_crypto.c

```diff
@@ -2840,6 +2842,260 @@ psa_status_t psa_mac_verify_finish( psa_mac_operation_t *operation,
 /* Asymmetric cryptography */
 /****************************************************************/
 
+typedef enum
+{
+    PSA_SIGN_INVALID = 0,
+    PSA_SIGN_HASH = 1,
+    PSA_SIGN_MESSAGE
+} psa_sign_operation_t;
+
+typedef enum
+{
+    PSA_VERIFY_INVALID = 0,
+    PSA_VERIFY_HASH = 1,
+    PSA_VERIFY_MESSAGE
+} psa_verify_operation_t;
+
+static psa_status_t psa_sign_internal( mbedtls_svc_key_id_t key,
+                                       psa_sign_operation_t operation,
+                                       psa_algorithm_t alg,
+                                       const uint8_t * input,
+                                       size_t input_length,
+                                       uint8_t * signature,
+                                       size_t signature_size,
+                                       size_t * signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    *signature_length = 0;
+
+    if( operation == PSA_SIGN_MESSAGE )
+    {
+        if( ! PSA_ALG_IS_SIGN_MESSAGE( alg ) )
+            return( PSA_ERROR_INVALID_ARGUMENT );
+
+        if ( PSA_ALG_IS_HASH_AND_SIGN( alg ) )
+        {
+            if( ! PSA_ALG_IS_HASH( PSA_ALG_SIGN_GET_HASH( alg ) ) )
+                return( PSA_ERROR_INVALID_ARGUMENT );
+        }
+        /* Curently only hash-then-sign algorithms are supported. */
+        else
+            return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    else if( operation == PSA_SIGN_INVALID )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
```

Comment:
Is there a missing case here for algorithm validation on `operation == PSA_SIGN_HASH`?